### PR TITLE
patchelf: conan v2

### PR DESCRIPTION
**Pull request by connorsmacd**
Specify library name and version:  **patchelf/***

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/patchelf/all/conanfile.py [hunk 1]
```diff
@@ -1,8 +1,16 @@
...
+from conan.tools.layout import basic_layout
+from conan.tools.scm import Version
+
 import os
 
-required_conan_version = ">=1.33.0"
+required_conan_version = ">=1.54.0"
 
 class PatchElfConan(ConanFile):
     name = "patchelf"
```
Comment (StellaSmith) [inline recipes/patchelf/all/conanfile.py:16]:
```suggestion
    name = "patchelf"
    package_type = "application"
```

**Review by StellaSmith — CHANGES_REQUESTED** [review]

**Review by StellaSmith — APPROVED** [review]

File: recipes/patchelf/all/test_package/conanfile.py [hunk 2]
```diff
@@ -1,9 +1,15 @@
...
+    generators = "VirtualBuildEnv"
+    test_type = "explicit"
+
+    def build_requirements(self):
+        self.tool_requires(self.tested_reference_str)
 
     def test(self):
-        if not tools.cross_building(self, skip_x64_x86=True):
-            self.run("patchelf --version", run_environment=True)
+        if not cross_building(self, skip_x64_x86=True):
```
Comment (prince-chrismc) [inline recipes/patchelf/all/test_package/conanfile.py:14]: There's a new better helper called `can_run` for this

File: recipes/patchelf/all/conanfile.py [hunk 3]
```diff
@@ -1,57 +1,79 @@
...
-        with tools.chdir(self._source_subfolder):
-            self.run("{} -fiv --warnings=all".format(tools.get_env("AUTORECONF")), run_environment=True)
-        autotools = self._configure_autotools()
+        autotools = Autotools(self)
+        autotools.autoreconf()
+        autotools.configure()
         autotools.make()
 
     def package(self):
-        self.copy(pattern="COPYING", src=self._source_subfolder, dst="licenses")
```
[the remark this answers is not in the document]
Comment (connorsmacd) [reply recipes/patchelf/all/conanfile.py:46]: I assumed the sources never contained a license file to begin with. If the file was removed at some point, wouldn't that cause the download to fail due to a SHA-256 mismatch?

I'm a new contributor here, so I don't have the full context. Perhaps this recipe was merged before the license check was in place, so no one ever noticed?

[the remark this answers is not in the document]
Comment (prince-chrismc) [reply recipes/patchelf/all/conanfile.py:46]:
```
[HOOK - conan-center.py] post_package(): ERROR: [PACKAGE LICENSE (KB-H012)] No 'licenses' folder found in package: /home/conan/w/prod/BuildSingleReference/.conan/data/patchelf/0.13/_/_/package/24647d9fe8ec489125dfbae4b3ebefaf7581674c  (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H012-PACKAGE-LICENSE) 
ERROR: [HOOK - conan-center.py] post_package(): Some checks failed running the hook, check the output
```

https://github.com/conan-io/conan-center-index/pull/16501#issuecomment-1463958680

[the remark this answers is not in the document]
Comment (prince-chrismc) [reply recipes/patchelf/all/conanfile.py:46]: https://github.com/conan-io/conan-center-index/blob/118bd0e9bb3a3fff80919e3748cf2cabe80bd48d/recipes/patchelf/all/conanfile.py#L57 🤔 

That looks right.... hmm 🧐

[the remark this answers is not in the document]
Comment (prince-chrismc) [reply recipes/patchelf/all/conanfile.py:46]: Ohhhhhh... the file was `"COPYING"` not `"LICENSE"` I think you made a copy paste mistake :)

File: recipes/patchelf/all/conanfile.py [hunk 4]
```diff
@@ -1,57 +1,79 @@
...
     def validate(self):
-        if not tools.is_apple_os(self.settings.os) and self.settings.os not in ("FreeBSD", "Linux"):
+        if not is_apple_os(self) and self.settings.os not in ("FreeBSD", "Linux"):
             raise ConanInvalidConfiguration("PatchELF is only available for GNU-like operating systems (e.g. Linux)")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version], destination=self._source_subfolder, strip_root=True)
+        get(self, **self.conan_data["sources"][self.version][0], strip_root=True)
+        # The library sources do not contain a license file, so download it from elsewhere
+        download(self, filename="LICENSE", **self.conan_data["sources"][self.version][1])
```
Comment (samuel-emrys) [inline recipes/patchelf/all/conanfile.py:38]: The sources do contain the license. The license is titled `COPYING`. I have verified by inspection of the source archive.
```suggestion
```

File: recipes/patchelf/all/conandata.yml [hunk 5]
```diff
@@ -1,7 +1,11 @@
 sources:
   "0.13":
-    url: "https://github.com/NixOS/patchelf/archive/0.13.tar.gz"
-    sha256: "60c6aeadb673de9cc1838b630c81f61e31c501de324ef7f1e8094a2431197d09"
+    - url: "https://github.com/NixOS/patchelf/archive/0.13.tar.gz"
+      sha256: "60c6aeadb673de9cc1838b630c81f61e31c501de324ef7f1e8094a2431197d09"
+    - url: "https://gnu.org/licenses/gpl-3.0.txt"
+      sha256: "3972dc9744f6499f0f9b2dbf76696f2ae7ad8af9b23dde66d6af86c9dfb36986"
```
Comment (samuel-emrys) [inline recipes/patchelf/all/conandata.yml:6]:
```suggestion
```

File: recipes/patchelf/all/conandata.yml [hunk 6]
```diff
@@ -1,7 +1,11 @@
...
   "0.12":
-    url: "https://github.com/NixOS/patchelf/archive/refs/tags/0.12.tar.gz"
-    sha256: "3dca33fb862213b3541350e1da262249959595903f559eae0fbc68966e9c3f56"
+    - url: "https://github.com/NixOS/patchelf/archive/refs/tags/0.12.tar.gz"
+      sha256: "3dca33fb862213b3541350e1da262249959595903f559eae0fbc68966e9c3f56"
+    - url: "https://gnu.org/licenses/gpl-3.0.txt"
+      sha256: "3972dc9744f6499f0f9b2dbf76696f2ae7ad8af9b23dde66d6af86c9dfb36986"
```
Comment (samuel-emrys) [inline recipes/patchelf/all/conandata.yml:11]:
```suggestion
```

File: recipes/patchelf/all/conanfile.py [hunk 7]
```diff
@@ -1,57 +1,79 @@
...
-        autotools = self._configure_autotools()
+        autotools = Autotools(self)
+        autotools.autoreconf()
+        autotools.configure()
         autotools.make()
 
     def package(self):
-        self.copy(pattern="COPYING", src=self._source_subfolder, dst="licenses")
-        autotools = self._configure_autotools()
+        copy(self, pattern="LICENSE", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
```
Comment (samuel-emrys) [inline recipes/patchelf/all/conanfile.py:61]: As per previous discussion
```suggestion
        copy(self, pattern="COPYING", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
```

File: recipes/patchelf/all/test_package/conanfile.py [hunk 8]
```diff
@@ -1,9 +1,15 @@
...
+            self.run("patchelf --version")
```
Comment (samuel-emrys) [inline recipes/patchelf/all/test_package/conanfile.py:15]: Could we get a more meaningful test of the executable here? This tests that it's on the path but not that it works - I've had hit and miss experience with `--version`. Often it doesn't exercise whether the executable has been packaged correctly.

See:
* https://github.com/conan-io/conan-center-index/pull/16276#discussion_r1117892744
* https://github.com/conan-io/conan-center-index/issues/14120

**Review by samuel-emrys — CHANGES_REQUESTED** [review]

**Review by SSE4 — DISMISSED** [review]

File: recipes/patchelf/all/test_package/conanfile.py (hunk 8 above)
Comment (mayeut) [reply recipes/patchelf/all/test_package/conanfile.py:15]: `patchelf` is self contained. There's no other files packaged than this single executable and the license so, from a packaging standpoint, `--version` is enough IMHO.

File: recipes/patchelf/all/conanfile.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (connorsmacd) [reply recipes/patchelf/all/conanfile.py:46]: Ah, my bad!

**Review by prince-chrismc — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

**Pull request merged**